### PR TITLE
Add nearest predicate to ArborX wrapper

### PR DESCRIPTION
**Pull request by Rombur**
This PR adds a nearest neighbor predicate to ArborX wrappers. While adding the predicate, I realized that the current design was not extensible and that the names were not great. I have created base classes that contains most of the implementation and added `Predicate` to the name of the classes since that's what they are. The classes were introduces a month ago so there is no backward compatibility problem.

### CODE REVIEW

File: doc/news/changes/minor/20210319Turcksin [hunk 1]
```diff
@@ -0,0 +1,5 @@
+New: Add nearest neighbor predicate support to ArborXWrappers::BVH(). This
+allows to find the n closest bounding boxes or points to any given bounding
```
Comment (jppelteret) [inline doc/news/changes/minor/20210319Turcksin:2]:
```suggestion
allows to find the `n` closest bounding boxes or points to any given bounding
```

File: include/deal.II/arborx/access_traits.h [hunk 2]
```diff
@@ -61,19 +59,68 @@ namespace ArborXWrappers
...
+
+
+  /**
+   * Base class for BoundingBox predicates.
+   */
+  class BoundingBoxPredicate
+  {
+  public:
+    /**
+     * Constructor. @p bounding_boes is a list of bounding boxes used by the
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:118]:
```suggestion
     * Constructor. @p bounding_boxes is a list of bounding boxes used by the
```

File: include/deal.II/arborx/bvh.h [hunk 3]
```diff
@@ -60,6 +60,12 @@ namespace ArborXWrappers
     template <int dim, typename Number>
     BVH(const std::vector<BoundingBox<dim, Number>> &bounding_boxes);
 
+    /**
+     * Constructor. Use a vector of Points @p points as primitives.
```
Comment (jppelteret) [inline include/deal.II/arborx/bvh.h:64]:
```suggestion
     * Constructor. Use a vector of @p points as primitives.
```

**Review by jppelteret — CHANGES_REQUESTED** [review]
A couple of superficial comments, but otherwise LGTM!

**Comment by Rombur** [comment]
Done.

File: include/deal.II/arborx/access_traits.h [hunk 4]
```diff
@@ -61,19 +59,68 @@ namespace ArborXWrappers
...
-      const std::vector<dealii::BoundingBox<dim, Number>> &bb);
+    PointIntersectPredicate(
+      const std::vector<dealii::Point<dim, Number>> &points);
+  };
+
+
+
+  /**
+   * This class defines a predicate used by ArborXWrappers::BVH to determine
+   * which points are the nearesst to the bounding boxes/points used to build
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:82]:
```suggestion
   * which points are the nearest to the bounding boxes/points used to build
```

File: include/deal.II/arborx/access_traits.h [hunk 5]
```diff
@@ -149,35 +245,62 @@ namespace ArborX
...
+    /**
+     * Number of Point stored in @p pt_nearest.
+     */
+    static std::size_t
+    size(const dealii::ArborXWrappers::PointNearestPredicate &pt_nearest);
+
+    /**
+     * Return an Arbox::nearest(ArborX::Point,
+     * PointNearestPredicate::get_n_nearest_neighbors) object constructed from
+     * the `i`th dealii::Point stored in @p pt_nearest
```
Comment (jppelteret) [inline include/deal.II/arborx/access_traits.h:290]:
```suggestion
     * the `i`th dealii::Point stored in @p pt_nearest.
```

**Review by jppelteret — APPROVED** [review]
Hi @Rombur, I've found another couple of minor typos. I'm sorry that I didn't spot them earlier. Perhaps you can fix them up, and I'll approve in the mean time to minimise the inconvenience.

File: include/deal.II/arborx/access_traits.h [hunk 6]
```diff
@@ -90,6 +137,55 @@ namespace ArborXWrappers
   private:
     std::vector<dealii::BoundingBox<3, float>> bounding_boxes;
   };
+
+
+
+  /**
+   *  This class is used by ArborXWrappers::BVH to determine which bounding
+   *  boxes intersect the bounding boxes used to build the ArborXWrappers::BVH.
```
Comment (masterleinad) [inline include/deal.II/arborx/access_traits.h:145]:
```suggestion
   *  This class is used by ArborXWrappers::BVH to determine for given bounding boxes
   *  which of the bounding boxes used to build the ArborXWrappers::BVH intersect with them.
```

File: include/deal.II/arborx/access_traits.h [hunk 7]
```diff
@@ -61,19 +59,68 @@ namespace ArborXWrappers
 
 
   /**
-   *  This class is used by ArborXWrappers::BVH to determine which bounding
-   *  boxes intersect the bounding boxes used to build the ArborXWrappers::BVH.
+   * This class defines a predicate used by ArborXWrappers::BVH to determine
+   * which points intersect the bounding boxes used to build the
+   * ArborXWrappers::BVH.
```
Comment (masterleinad) [inline include/deal.II/arborx/access_traits.h:64]:
```suggestion
   * This class defines a predicate used by ArborXWrappers::BVH to determine
   * for given points which of the bounding boxes used to build the
   * ArborXWrappers::BVH intersect with them.
```

File: include/deal.II/arborx/access_traits.h [hunk 8]
```diff
@@ -90,6 +137,55 @@ namespace ArborXWrappers
...
+   */
+  class BoundingBoxNearestPredicate : public BoundingBoxPredicate
+  {
+  public:
+    /**
+     * Constructor. @p bounding_boxes is a list of bounding boxes which we are interested in
+     * knowing if they are the nearest to the ArborXWrappers::BVH bounding
+     * boxes.
+     * @p n_nearest_neighbors is the number of nearest neighbors we are looking
+     * for.
```
Comment (masterleinad) [inline include/deal.II/arborx/access_traits.h:173]:
```suggestion
     * Constructor. @p bounding_boxes is a list of bounding boxes for which are interested in
     * knowing the @p n_nearest_neighbors nearest bounding boxes used to build the
     * ArborXWrappers::BVH.
```

**Comment by jppelteret** [comment]
Great, thanks a lot!

**Comment by jppelteret** [comment]
I see that @masterleinad wants to review this as well, so let's give him some time to take a look at it.

**Review by masterleinad — COMMENTED** [review]
I think the comments are phrased backward. Otherwise, this looks good to me.

**Comment by Rombur** [comment]
@masterleinad You are right. Thanks. I have fixed it.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by bangerth** [comment]
There are two approvals of this PR, so I'm not going to take a third view in detail. But I do want to ask: Base classes are generally used as "interfaces" with `virtual` functions. But yours don't have virtual functions; you might want to elaborate on how derived classes are supposed to use these classes as "base classes".

(A related example is `Quadrature`, which is also a base class without virtual functions.)

**Comment by jppelteret** [comment]
@bangerth's comment got me thinking... do we at least need to make the destructor for the base class `PointPredicate` virtual?

**Comment by masterleinad** [comment]
I think we just want to use a base class to share code the derived classes have in common. We don't overwrite any member functions and thus they don't have to be virtual. Since none of the derived classes does anything meaningful in their destructor, having the destructor `virtual` is not crucial but maybe good practice.

**Comment by Rombur** [comment]
Oh sorry, I didn't these messages. Like @masterleinad said the base class is not an interface. Because of the design of ArborX, I need two different class that have the same implementation but one of the two needs to store an extra variable. I moved all the implementation into the base class and I derived two classes but they are empty shells.

**Comment by masterleinad** [comment]
That being said we could just use `private` or `protected` inheritance to make sure that we don't use them in a polymorphic context.

**Comment by Rombur** [comment]
@masterleinad that's a good idea

**Pull request merged**